receiver/wavefrontreceiver/wavefront_parser.go

```diff
@@ -0,0 +1,274 @@
+		if strings.IndexByte(timestampStr, '=') == -1 {
+			return nil, fmt.Errorf(
+				"invalid timestamp for wavefront metric [%s]", line)
+		}
+		// Assume timestamp was omitted, get current time and adjust index.
+		ts.Seconds = time.Now().Unix()
+		tags = rest
+	}
+	point.Timestamp = &ts
+
+	var labelKeys []*metricspb.LabelKey
+	var labelValues []*metricspb.LabelValue
+	if tags != "" {
+		// to need for special treatment for source, treat it as a normal tag since
+		// tags are separated by space and are optionally double-quoted.
+		var err error
+		labelKeys, labelValues, err = buildLabels(tags)
+		if err != nil {
+			return nil, fmt.Errorf("invalid wavefront metric [%s]: %v", line, err)
+		}
+	}
+
+	if wp.ExtractCollectdTags {
+		metricName, labelKeys, labelValues = wp.injectCollectDLabels(metricName, labelKeys, labelValues)
+	}
+
+	metric := &metricspb.Metric{
+		MetricDescriptor: &metricspb.MetricDescriptor{
+			Name:      metricName,
+			Type:      metricType,
+			LabelKeys: labelKeys,
+		},
+		Timeseries: []*metricspb.TimeSeries{
+			{
+				LabelValues: labelValues,
+				Points:      []*metricspb.Point{&point},
+			},
+		},
+	}
+	return metric, nil
+}
+
+func (wp *WavefrontParser) injectCollectDLabels(
+	metricName string,
+	labelKeys []*metricspb.LabelKey,
+	labelValues []*metricspb.LabelValue,
+) (string, []*metricspb.LabelKey, []*metricspb.LabelValue) {
+	// This comes from SignalFx Gateway code that has the capability to
+	// remove CollectD tags from the name of the metric.
+	var toAddDims map[string]string
+	index := strings.Index(metricName, "..")
+	for {
+		metricName, toAddDims = labelsFromName(&metricName)
+		if len(toAddDims) == 0 {
+			if index == -1 {
+				metricName = strings.Replace(metricName, "..", ".", -1)
+			}
+
+			break
+		}
+
+		for k, v := range toAddDims {
+			labelKeys = append(labelKeys, &metricspb.LabelKey{Key: k})
+			labelValues = append(labelValues, &metricspb.LabelValue{
+				Value:    v,
+				HasValue: true,
+			})
+		}
+	}
+	return metricName, labelKeys, labelValues
+}
+
+func buildLabels(tags string) (keys []*metricspb.LabelKey, values []*metricspb.LabelValue, err error) {
+	if tags == "" {
+		return
+	}
+	for {
+		parts := strings.SplitN(tags, "=", 2)
+		if len(parts) != 2 {
+			return nil, nil, fmt.Errorf("failed to break key for [%s]", tags)
+		}
+
+		key := parts[0]
+		rest := parts[1]
+		tagLen := len(key) + 1 // Length of key plus separator and yet to be determined length of the value.
+		var value string
+		if len(rest) > 1 && rest[0] == '"' {
+			// Skip until non-escaped double quote.
+			foundEscape := false
+			i := 1
+			for ; i < len(rest); i++ {
+				if rest[i] != '"' {
+					continue
+				}
+				if rest[i-1] != '\\' {
+					break
+				}
+				foundEscape = true
+			}
+
+			value = rest[1:i]
+			tagLen += len(value) + 2 // plus 2 to account for the double-quotes.
+			if foundEscape {
+				value = escapedDoubleQuoteReplacer.Replace(value)
```

Comment:
Reading the wavefront spec it does not explicitly say that backslash can precede only double quote character. I think it does not exclude the possibility that anything can be escaped by a backslash. Should backslash escaping be handled more generally and not just double quotes?

Reply:
Agreed: the spec is a bit ambiguous in this regard. However, I looked at the wavefront sdk and it just escapes the double-quotes, see https://github.com/wavefrontHQ/wavefront-sdk-go/blob/2c5891318fcd83c35c93bba2b411640495473333/senders/formatter.go#L364 I should have added a link to that as refernce.

The Gateway actually passes the string as it received but since we don't know the destination on the Collector we need to let the exporters do the escaping that applies to each one.

Reply:
Good that you made me look at it again: a missed the escaping of line-breaks...